collectors/perf.plugin/perf_plugin.c

```diff
@@ -0,0 +1,1230 @@
+#define NETDATA_CHART_PRIO_PERF_MIGRATIONS            8810
+#define NETDATA_CHART_PRIO_PERF_ALIGNMENT             8811
+#define NETDATA_CHART_PRIO_PERF_EMULATION             8812
+
+// Hardware cache counters
+#define NETDATA_CHART_PRIO_PERF_L1D                   8820
+#define NETDATA_CHART_PRIO_PERF_L1D_PREFETCH          8821
+#define NETDATA_CHART_PRIO_PERF_L1I                   8822
+#define NETDATA_CHART_PRIO_PERF_LL                    8823
+#define NETDATA_CHART_PRIO_PERF_DTLB                  8824
+#define NETDATA_CHART_PRIO_PERF_ITLB                  8825
+#define NETDATA_CHART_PRIO_PERF_PBU                   8826
+
+// callback required by fatal()
+void netdata_cleanup_and_exit(int ret) {
+    exit(ret);
+}
+
+void send_statistics( const char *action, const char *action_result, const char *action_data) {
+    (void) action;
+    (void) action_result;
+    (void) action_data;
+    return;
+}
+
+// callbacks required by popen()
+void signals_block(void) {};
+void signals_unblock(void) {};
+void signals_reset(void) {};
+
+// callback required by eval()
+int health_variable_lookup(const char *variable, uint32_t hash, struct rrdcalc *rc, calculated_number *result) {
+    (void)variable;
+    (void)hash;
+    (void)rc;
+    (void)result;
+    return 0;
+};
+
+// required by get_system_cpus()
+char *netdata_configured_host_prefix = "";
+
+// Variables
+
+#define RRD_TYPE_PERF "perf"
+#define RRD_FAMILY_HW "hardware"
+#define RRD_FAMILY_SW "software"
+#define RRD_FAMILY_CACHE "cache"
+
+#define NO_FD -1
+#define ALL_PIDS -1
+#define UINT64_SIZE 8
+
+static int debug = 0;
+
+static int update_every = 1;
+static int freq = 0;
+
+typedef enum perf_event_id {
+    // Hardware counters
+    EV_ID_CPU_CYCLES,
+    EV_ID_INSTRUCTIONS,
+    EV_ID_CACHE_REFERENCES,
+    EV_ID_CACHE_MISSES,
+    EV_ID_BRANCH_INSTRUCTIONS,
+    EV_ID_BRANCH_MISSES,
+    EV_ID_BUS_CYCLES,
+    EV_ID_STALLED_CYCLES_FRONTEND,
+    EV_ID_STALLED_CYCLES_BACKEND,
+    EV_ID_REF_CPU_CYCLES,
+
+    // Software counters
+    // EV_ID_CPU_CLOCK,
+    // EV_ID_TASK_CLOCK,
+    // EV_ID_PAGE_FAULTS,
+    // EV_ID_CONTEXT_SWITCHES,
+    EV_ID_CPU_MIGRATIONS,
+    // EV_ID_PAGE_FAULTS_MIN,
+    // EV_ID_PAGE_FAULTS_MAJ,
+    EV_ID_ALIGNMENT_FAULTS,
+    EV_ID_EMULATION_FAULTS,
+
+    // Hardware cache counters
+    EV_ID_L1D_READ_ACCESS,
+    EV_ID_L1D_READ_MISS,
+    EV_ID_L1D_WRITE_ACCESS,
+    EV_ID_L1D_WRITE_MISS,
+    EV_ID_L1D_PREFETCH_ACCESS,
+
+    EV_ID_L1I_READ_ACCESS,
+    EV_ID_L1I_READ_MISS,
+
+    EV_ID_LL_READ_ACCESS,
+    EV_ID_LL_READ_MISS,
+    EV_ID_LL_WRITE_ACCESS,
+    EV_ID_LL_WRITE_MISS,
+
+    EV_ID_DTLB_READ_ACCESS,
+    EV_ID_DTLB_READ_MISS,
+    EV_ID_DTLB_WRITE_ACCESS,
+    EV_ID_DTLB_WRITE_MISS,
+
+    EV_ID_ITLB_READ_ACCESS,
+    EV_ID_ITLB_READ_MISS,
+
+    EV_ID_PBU_READ_ACCESS,
+
+    EV_ID_END
+} perf_event_id_t;
+
+enum perf_event_group {
+    EV_GROUP_0,
+    EV_GROUP_1,
+    EV_GROUP_2,
+    EV_GROUP_3,
+    EV_GROUP_4,
+    EV_GROUP_5,
```

Comment:
Do these groups have any meaning? If so i think it is better to decode meaning in the name instead of using number.

Reply:
Renamed